File: ios/ConverseNotificationExtension/NotificationHandler.swift
//
//  NotificationHandler.swift
//  ConverseNotificationExtension
//
//  Created by lourou on 2023-10-13.
//

import Foundation
import MMKVAppExtension

func incrementBadge(for content: UNMutableNotificationContent) {
  let newBadgeCount = getBadge() + 1
  setBadge(newBadgeCount)
  content.badge = NSNumber(value: newBadgeCount)
}

func shouldShowNotification(for messageId: String?) -> Bool {
  let maxStoredIds = 10
  
  // Check if the messageId is not nil and not an empty string
  guard let id = messageId, !id.isEmpty else {
    return false
  }
  
  // If the id already exists in the list, don't show the notification
  let existingIds = getShownNotificationIds()
  if existingIds.contains(id) {
    return false
  }

  // Append the new id to the list
  var updatedIds = existingIds
  updatedIds.append(id)

  // If the list exceeds the maximum limit, trim from the beginning
  if updatedIds.count > maxStoredIds {
    updatedIds.removeFirst(updatedIds.count - maxStoredIds)
  }

  // Store the updated list of IDs back to storage
  setShownNotificationIds(updatedIds)
  
  // If all conditions are met, show the notification
  return true
}

File: android/app/src/main/java/com/converse/dev/NotificationHandler.kt
package com.converse.dev

import android.util.Log
import android.content.Context

class NotificationHandler {
    companion object {
        private const val TAG = "NotificationHandler"
        private const val MAX_STORED_IDS = 10
    }

    fun notificationAlreadyShown(appContext: Context, messageId: String?): Boolean {
        Log.d(TAG, "Checking if we should show notification for message ID: $messageId")

        messageId?.let {
            val existingIds = getShownNotificationIds(appContext)

            if (existingIds.contains(it)) {
                Log.d(TAG, "Message ID already exists. Not showing notification.")
                return true
            }

            val updatedIds = existingIds + it
            val trimmedList = if (updatedIds.size > MAX_STORED_IDS) {
                updatedIds.subList(updatedIds.size - MAX_STORED_IDS, updatedIds.size)
            } else {
                updatedIds
            }

            setShownNotificationIds(appContext, trimmedList)

            val storedData = getShownNotificationIds(appContext).joinToString(", ")
            Log.d(TAG, "Stored IDs just after saving: $storedData")

            return false
        }
        Log.d(TAG, "Message ID is null. Not showing notification.")
        return true
    }
}